### PR TITLE
Crash USI stream on Mesos master connection loss.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
This patch introduces a simple integration test that verifies the USI stream closes if the connection to Mesos is lost.

JIRA issues: D2IQ-67087

### CODE REVIEW

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 13 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 18 seconds.
[info] Total number of tests run: 73
[info] Suites: completed 11, aborted 0
[info] Tests: succeeded 72, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	com.mesosphere.usi.core.integration.SchedulerIntegrationTest
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 104 s (01:44), completed Apr 17, 2020 2:50:54 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by jeschkies** [comment]
So I see the log line from https://github.com/mesosphere/usi/blob/master/core/src/main/scala/com/mesosphere/usi/core/SchedulerFactory.scala#L97:
```
Mesos client hanging up due to error in stream
```

**Comment by jeschkies** [comment]
Call test with
```
sbt 'core/testOnly *SchedulerIntegrationTest -- -z "terminates the stream when Mesos master dies"'
```

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[error] 	com.mesosphere.usi.core.integration.SchedulerIntegrationTest
[info] Run completed in 1 minute, 17 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 24 seconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 15 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 103 s (01:43), completed Apr 17, 2020 3:13:50 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Review by timcharper — APPROVED** [review]

**Comment by jeschkies** [comment]
@timcharper, this is not a fix. This is just a reproduction of the issue.

**Review by kaiwalyajoshi — APPROVED** [review]

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 10 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 15 seconds.
[info] Total number of tests run: 73
[info] Suites: completed 11, aborted 0
[info] Tests: succeeded 72, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	com.mesosphere.usi.core.integration.SchedulerIntegrationTest
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 98 s (01:38), completed Apr 23, 2020 8:27:17 AM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by jeschkies** [comment]
I addes some logs and have these results
```
 === Test: terminates the stream when Mesos master dies ===
 ===== Given A Mesos client (SchedulerIntegrationTest.scala:115) =====
 ===== And a first successful command. (SchedulerIntegrationTest.scala:128) =====
Connecting to Mesos master http://192.168.1.105:32772
 ===== When Mesos masters crash. (SchedulerIntegrationTest.scala:135) =====
Kill Mesos masters.
===== Then The next command should fail (SchedulerIntegrationTest.scala:139) =====
Offer KillPod returned Enqueued
 ===== And The stream terminates (SchedulerIntegrationTest.scala:145) =====
Received USI event Some(PodSpecUpdatedEvent(PodId(unknown-pod-1),None))
```

So, the second offer should have failed and it is very odd that we receive a `PodSpecUpdatedEvent(PodId(unknown-pod-1),None)` even though we've sent `KillPod(PodId("unknown-pod-2"))`.

**Comment by timcharper** [comment]
Exit status 1 when running "sbt scalafmtCheck test:scalafmtCheck"

<details>
<summary>Details</summary>
<pre>
-Xmx2048M
-Xss6M
-XX:MaxPermSize=512M
-jar
/home/travis/.sbt/launchers/1.3.3/sbt-launch.jar
scalafmtCheck
test:scalafmtCheck

saving stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[info] Loading settings for project usi-build from plugins.sbt ...
[info] Loading project definition from /home/travis/build/mesosphere/usi/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/travis/build/mesosphere/usi/)
[success] Total time: 7 s, completed Apr 23, 2020 8:58:47 AM
[error] /home/travis/build/mesosphere/usi/core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala isn&#39;t formatted properly!
[error] (core / Test / scalafmtCheck) /home/travis/build/mesosphere/usi/core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala isn&#39;t formatted properly!
[error] Total time: 1 s, completed Apr 23, 2020 8:58:49 AM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[info] No tests were executed.
[info] Run completed in 1 minute, 19 seconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 30 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 16 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 104 s (01:44), completed Apr 23, 2020 9:01:12 AM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "sbt scalafmtCheck test:scalafmtCheck"

<details>
<summary>Details</summary>
<pre>
-Xms2048M
-Xmx2048M
-Xss6M
-XX:MaxPermSize=512M
-jar
/home/travis/.sbt/launchers/1.3.3/sbt-launch.jar
scalafmtCheck
test:scalafmtCheck

saving stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[info] Loading settings for project usi-build from plugins.sbt ...
[info] Loading project definition from /home/travis/build/mesosphere/usi/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/travis/build/mesosphere/usi/)
[error] /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] (core / Compile / scalafmtCheck) /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] Total time: 6 s, completed Apr 23, 2020 2:27:56 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[info] All tests passed.
[info] Run completed in 1 minute, 30 seconds.
[info] Total number of tests run: 11
[info] Suites: completed 3, aborted 0
[info] Tests: succeeded 11, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 18 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 11 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 104 s (01:44), completed Apr 23, 2020 2:30:22 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

File: core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala [hunk 1]
```diff
@@ -110,4 +110,45 @@ class SchedulerIntegrationTest extends AkkaUnitTest with MesosClusterTest with I
...
+      case PodSpecUpdatedEvent(PodId(idString), _) =>
+        idString should be("unknown-pod-1")
+    }
+
+    When("Mesos masters crash.")
+    logger.info("Kill Mesos masters.")
+    mesosCluster.masters.foreach(_.stop())
+
+    Then("The stream terminates")
+    val completion = input.watchCompletion().futureValue
```
Comment (jeschkies) [inline core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala]: The logs state
```
[scheduler commands] Downstream finished, cause: EntityStreamException: Entity stream truncatio    582 n. The HTTP parser was receiving an entity when the underlying connection was closed unexpectedly.
```
But the `watchComplete` returns a successful future with `Done`. Why is that, @timcharper?

File: core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala [hunk 2]
```diff
@@ -110,4 +110,45 @@ class SchedulerIntegrationTest extends AkkaUnitTest with MesosClusterTest with I
...
+    //completion.getMessage should be("I failed")
+
+    And("The next command should fail")
+    val offerResult = input.offer(commands.KillPod(PodId("unknown-pod-2"))).failed.futureValue
+    //offerResult.getMessage should be("foot")
+//    inside(offerResult) {
+//      case QueueOfferResult.Failure(cause) => cause.getMessage should be("foo")
+//    }
+
+    val pullResult = output.pull().failed.futureValue
```
Comment (jeschkies) [inline core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala]: The pull does not fail, why @timcharper? It just returns `None`.

**Comment by timcharper** [comment]
Exit status 1 when running "sbt scalafmtCheck test:scalafmtCheck"

<details>
<summary>Details</summary>
<pre>
-Xms2048M
-Xmx2048M
-Xss6M
-XX:MaxPermSize=512M
-jar
/home/travis/.sbt/launchers/1.3.3/sbt-launch.jar
scalafmtCheck
test:scalafmtCheck

saving stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[info] Loading settings for project usi-build from plugins.sbt ...
[info] Loading project definition from /home/travis/build/mesosphere/usi/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/travis/build/mesosphere/usi/)
[error] /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] (core / Compile / scalafmtCheck) /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] Total time: 5 s, completed Apr 23, 2020 4:09:48 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Unused imports / scalafix"

<details>
<summary>Details</summary>
<pre>
 import akka.event.Logging
 import akka.stream.scaladsl.{Keep, Sink}
-import akka.stream.{ActorMaterializer, Attributes, QueueOfferResult}
+import akka.stream.{ActorMaterializer, Attributes}
 import com.mesosphere.mesos.client.MesosClient
 import com.mesosphere.mesos.conf.MesosClientSettings
 import com.mesosphere.usi.core.SchedulerFactory
@@ -142,7 +142,7 @@
     //completion.getMessage should be(&quot;I failed&quot;)
 
     And(&quot;The next command should fail&quot;)
-    val offerResult = input.offer(commands.KillPod(PodId(&quot;unknown-pod-2&quot;))).failed.futureValue
+    input.offer(commands.KillPod(PodId(&quot;unknown-pod-2&quot;))).failed.futureValue
     //offerResult.getMessage should be(&quot;foot&quot;)
 //    inside(offerResult) {
 //      case QueueOfferResult.Failure(cause) =&gt; cause.getMessage should be(&quot;foo&quot;)
[error] (core / Test / scalafix) scalafix.sbt.ScalafixFailed: TestError
[error] Total time: 4 s, completed Apr 23, 2020 4:10:13 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Running tests for all Scala versions"

<details>
<summary>Details</summary>
<pre>
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[info] Run completed in 1 minute, 18 seconds.
[info] Total number of tests run: 73
[info] Suites: completed 11, aborted 0
[info] Tests: succeeded 72, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	com.mesosphere.usi.core.integration.SchedulerIntegrationTest
[info] Run completed in 1 minute, 16 seconds.
[info] Total number of tests run: 8
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 8, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[error] (core / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 101 s (01:41), completed Apr 23, 2020 4:12:08 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "sbt scalafmtCheck test:scalafmtCheck"

<details>
<summary>Details</summary>
<pre>
-Xms2048M
-Xmx2048M
-Xss6M
-XX:MaxPermSize=512M
-jar
/home/travis/.sbt/launchers/1.3.3/sbt-launch.jar
scalafmtCheck
test:scalafmtCheck

saving stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
OpenJDK 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
[info] Loading settings for project usi-build from plugins.sbt ...
[info] Loading project definition from /home/travis/build/mesosphere/usi/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/travis/build/mesosphere/usi/)
[error] /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] (core / Compile / scalafmtCheck) /home/travis/build/mesosphere/usi/core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala isn&#39;t formatted properly!
[error] Total time: 6 s, completed Apr 27, 2020 3:47:23 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Unused imports / scalafix"

<details>
<summary>Details</summary>
<pre>
[info] Running scalafix on 2 Scala sources
[info] Running scalafix on 3 Scala sources
[info] Running scalafix on 3 Scala sources
[info] Running scalafix on 14 Scala sources
[info] Running scalafix on 2 Scala sources
--- /home/travis/build/mesosphere/usi/core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala
+++ &lt;expected fix&gt;
@@ -2,7 +2,7 @@
 
 import akka.event.Logging
 import akka.stream.scaladsl.{Keep, Sink}
-import akka.stream.{ActorMaterializer, Attributes, QueueOfferResult}
+import akka.stream.{ActorMaterializer, Attributes}
 import com.mesosphere.mesos.client.MesosClient
 import com.mesosphere.mesos.conf.MesosClientSettings
 import com.mesosphere.usi.core.SchedulerFactory
[error] (core / Test / scalafix) scalafix.sbt.ScalafixFailed: TestError
[error] Total time: 4 s, completed Apr 27, 2020 3:47:48 PM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Comment by timcharper** [comment]
Exit status 1 when running "Unused imports / scalafix"

<details>
<summary>Details</summary>
<pre>
[info] Running scalafix on 3 Scala sources
[info] Running scalafix on 2 Scala sources
[info] Running scalafix on 3 Scala sources
[info] Running scalafix on 14 Scala sources
[info] Running scalafix on 2 Scala sources
--- /home/travis/build/mesosphere/usi/core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerIntegrationTest.scala
+++ &lt;expected fix&gt;
@@ -2,7 +2,7 @@
 
 import akka.event.Logging
 import akka.stream.scaladsl.{Keep, Sink}
-import akka.stream.{ActorMaterializer, Attributes, QueueOfferResult}
+import akka.stream.{ActorMaterializer, Attributes}
 import com.mesosphere.mesos.client.MesosClient
 import com.mesosphere.mesos.conf.MesosClientSettings
 import com.mesosphere.usi.core.SchedulerFactory
[error] (core / Test / scalafix) scalafix.sbt.ScalafixFailed: TestError
[error] Total time: 4 s, completed Apr 28, 2020 8:46:55 AM

restoring stty: 500:5:bf:8a3b:3:1c:7f:15:4:0:1:0:11:13:1a:0:12:f:17:16:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0:0
</pre>
</details>

**Pull request merged**